dali/ft/filecopy.cpp

```diff
@@ -2596,6 +2651,345 @@ void FileSprayer::pushParts()
+            case ApiCopyStatus::Pending:
+                try
+                {
+                    if (blobCopyOp->HasValue() && blobCopyOp->Value().CopyId.HasValue())
+                        blobClient->AbortCopyFromUri(blobCopyOp->Value().CopyId.Value().c_str());
+                    else // not sure that this can ever happen
+                        IERRLOG("Transfer using API .AbortCopy() failed: unexpected response from API");
+                    status = ApiCopyStatus::Aborted;
+                }
+                catch(const Azure::Core::RequestFailedException& e)
+                {
+                    IERRLOG("Abort copy operation failed: %s (code %d)", e.ReasonPhrase.c_str(), static_cast<int>(e.StatusCode));
+                    status = ApiCopyStatus::Failed;
+                }
+                break;
+            case ApiCopyStatus::Success:
+                // already copied-> need to delete
+                try
+                {
+                    blobClient->Delete();
+                }
+                catch(const Azure::Core::RequestFailedException& e)
+                {
+                    IERRLOG("Delete operation failed: %s (code %d)", e.ReasonPhrase.c_str(), static_cast<int>(e.StatusCode));
+                    status = ApiCopyStatus::Failed;
+                }
+                break;
+        }
+        return status;
+    }
+};
+
+void FileSprayer::transferUsingAPI()
+{
+    // N.B. Only call this member function, if FileSprayer::useAPICopy() has returned true
+    // Source
+    StringBuffer sourceClusterName, sourceBaseURI, sourceSAStoken;
+    distributedSource->getClusterName(0, sourceClusterName);
+    Owned<IStoragePlane> sourcePlane = getDataStoragePlane(sourceClusterName.str(), false);
+    CApiCopyClient::buildStorageURIBase(sourcePlane, sourceBaseURI);
+    sourcePlane->getSASToken(sourceSAStoken);
+
+    // Target
+    StringBuffer targetClusterName, targetBaseURI, targetSAStoken;
+    distributedTarget->getClusterName(0, targetClusterName);
+    Owned<IStoragePlane> targetPlane = getDataStoragePlane(targetClusterName.str(), false);
+    CApiCopyClient::buildStorageURIBase(targetPlane, targetBaseURI);
+    targetPlane->getSASToken(targetSAStoken);
+    StorageType targetType = targetPlane->getStorageType();
+
+    OwnedPointerArrayOf<IAPICopyClient> apiClients;
+    unsigned failCount = 0;
+    StringBuffer failedReason;
+    try
+    {
+        ForEachItemIn(idx1, partition)
+        {
+            PartitionPoint & cur = partition.item(idx1);
+            OutputProgress & curProgress = progress.item(idx1);
+            if (curProgress.status==OutputProgress::StatusCopied)
+            {
+                apiClients.append(nullptr); // represents parts already copied
+                continue;
+            }
+
+            StringBuffer sourceURI, targetURI;
+            CApiCopyClient::buildAzureStorageURI(distributedSource, cur.inputPartNum, sourceBaseURI.str(), sourceSAStoken.str(), sourceURI);
```

Comment:
You could have

`
inputPartNum = sources.item(cur.whichInput).partNum;
`
and avoid storing inputPartNum in the partition.

If it is used more than once you could add a

```
FileSprayer::getInputPartNum(unsigned whichInput)
{
   return sources.item(whichInput).partNum;
}

```